### PR TITLE
.*: Refine the checking of dag request

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: ref #7102

Problem Summary:

### What is changed and how it works?
- Wrapping `tipb::DAGRequest` as `DB::DAGRequest` to mask the difference between list based and tree based.
- Add checking of dagrequest and throw exception with badrequest err code.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/Coprocessor/DAGRequest.cpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        });
+    }
+    else
+    {
+        // generate executor_id for list based request.
+        ExecutorIdGenerator id_generator;
+        for (int i = 0; i < dag_request->executors_size(); ++i)
+        {
+            auto * executor = dag_request->mutable_executors(i);
+            check(!executor->has_executor_id(), fmt::format("for list based request, executor id must be null, the unexpected executor id: {}", executor->executor_id()));
```
Comment (ywqzzy) [inline dbms/src/Flash/Coprocessor/DAGRequest.cpp:109]: Not sure if it is safe.
```
// for disaggregated task on write node
DAGContext::DAGContext(tipb::DAGRequest & dag_request_, const disaggregated::DisaggTaskMeta & task_meta_, TablesRegionsInfo && tables_regions_info_, const String & compute_node_host_, LoggerPtr log_)
    : dag_request(&dag_request_)
    , dummy_query_string(dag_request->DebugString())
    , dummy_ast(makeDummyQuery())
    , tidb_host(compute_node_host_)
    , collect_execution_summaries(dag_request->has_collect_execution_summaries() && dag_request->collect_execution_summaries())
    , is_mpp_task(false)
    , is_root_mpp_task(false)
    , is_batch_cop(false)
    , is_disaggregated_task(true)
    , tables_regions_info(std::move(tables_regions_info_))
    , log(std::move(log_))
    , flags(dag_request->flags())
    , sql_mode(dag_request->sql_mode())
    , disaggregated_id(std::make_unique<DM::DisaggTaskId>(task_meta_))
    , max_recorded_error_count(getMaxErrorCount(*dag_request))
    , warnings(max_recorded_error_count)
    , warning_count(0)
    , keyspace_id(RequestUtils::deriveKeyspaceID(task_meta_))
{
    RUNTIME_CHECK(isTreeBasedExecutors() && dag_request->root_executor().has_executor_id());
    return_executor_id = dag_request->root_executor().has_executor_id() || dag_request->executors(0).has_executor_id();

    initOutputInfo();
    initListBasedExecutors();
}
```
It assumes that list-based executor may have `executor_id`

File: dbms/src/Flash/Coprocessor/DAGRequest.cpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+            executor->set_executor_id(executor_id);
+            // The begin of list_based_executors_order is the leaf node like table scan.
+            list_based_executors_order.push_back(executor_id);
+        }
+    }
+}
+
+const tipb::Executor & DAGRequest::rootExecutor() const
+{
+    check(dag_request, "dagrequest cannot be null");
```
Comment (ywqzzy) [inline dbms/src/Flash/Coprocessor/DAGRequest.cpp:122]:
```suggestion
    check(dag_request, "dagrequest can't be null");
```

File: dbms/src/Flash/Coprocessor/DAGRequest.cpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    return isTreeBased() ? dag_request->root_executor() : dag_request->executors(dag_request->executors_size() - 1);
+}
+
+void DAGRequest::traverse(std::function<bool(const tipb::Executor &)> && func) const
+{
+    check(dag_request, "dagrequest cannot be null");
```
Comment (ywqzzy) [inline dbms/src/Flash/Coprocessor/DAGRequest.cpp:128]:
```suggestion
    check(dag_request, "dagrequest can't be null");
```

File: dbms/src/Flash/Coprocessor/DAGRequest.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    traverseExecutors(dag_request, std::move(func));
+}
+
+void DAGRequest::traverseReverse(std::function<void(const tipb::Executor &)> && func) const
+{
+    check(dag_request, "dagrequest cannot be null");
```
Comment (ywqzzy) [inline dbms/src/Flash/Coprocessor/DAGRequest.cpp:134]:
```suggestion
    check(dag_request, "dagrequest can't be null");
```

File: dbms/src/Flash/Coprocessor/DAGRequest.cpp (hunk 1 above)
Comment (SeaRise) [reply dbms/src/Flash/Coprocessor/DAGRequest.cpp:109]: From https://github.com/pingcap/tidb/blob/3c942f7e101d6b0aff85c552598678f2a761deca/planner/core/plan_to_pb.go#L37 and https://github.com/pingcap/tispark/blob/c908c5516ea70c17b1bab5c0de3bafacb5b5c2a1/tikv-client/src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java#L234, executor_id will be empty in list base request.
But I think it's fine to not require the list based executor_id to be empty.
/cc @windtalker

Comment (ywqzzy) [reply dbms/src/Flash/Coprocessor/DAGRequest.cpp:109]: But IMO the request is send from TiFlash CN to DN, and we have changed the DAGRequest already.(I am not sure)

Comment (SeaRise) [reply dbms/src/Flash/Coprocessor/DAGRequest.cpp:109]: https://github.com/pingcap/tidb/blob/3c942f7e101d6b0aff85c552598678f2a761deca/planner/core/plan_to_pb.go#L160-L169
It seems to be a bug that constructs a list based request with an executor id.
Although this request will only be sent to tikv...

Comment (SeaRise) [reply dbms/src/Flash/Coprocessor/DAGRequest.cpp]: https://github.com/pingcap/tiflash/pull/7124/files#diff-759cb48a99e0dc5922fc34aa80a79f1ef4b99e6986deba493d993e59cbba99ffR29-R67
No longer assume that a list based executor must not have an executor id.

**Review by ywqzzy — APPROVED** [review]

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**